tests/test_list.h

```diff
@@ -45,6 +45,271 @@ static void populate_integers(List<int> &p_list, List<int>::Element *r_elements[
+	CHECK(n->get() == "C");
+
+	CHECK(list.size() == 3);
+	CHECK(!list.empty());
+
+	String v;
+	v = list.front()->get();
+	list.pop_front();
+	CHECK(v == "C");
+	v = list.front()->get();
+	list.pop_front();
+	CHECK(v == "B");
+	v = list.front()->get();
+	list.pop_front();
+	CHECK(v == "A");
+
+	CHECK(list.size() == 0);
+	CHECK(list.empty());
+
+	CHECK(list.back() == nullptr);
+	CHECK(list.front() == nullptr);
+}
+
+TEST_CASE("[List] Set and get") {
+	List<String> list;
+	list.push_back("A");
+
+	List<String>::Element *n = list.front();
+	CHECK(n->get() == "A");
+
+	n->set("X");
+	CHECK(n->get() == "X");
+}
+
+TEST_CASE("[List] Insert before") {
+	List<String> list;
+	List<String>::Element *a = list.push_back("A");
+	List<String>::Element *b = list.push_back("B");
+	List<String>::Element *c = list.push_back("C");
+
+	list.insert_before(b, "I");
+
+	CHECK(a->next()->get() == "I");
+	CHECK(c->prev()->prev()->get() == "I");
+	CHECK(list.front()->next()->get() == "I");
+	CHECK(list.back()->prev()->prev()->get() == "I");
+}
+
+TEST_CASE("[List] Insert after") {
+	List<String> list;
+	List<String>::Element *a = list.push_back("A");
+	List<String>::Element *b = list.push_back("B");
+	List<String>::Element *c = list.push_back("C");
+
+	list.insert_after(b, "I");
+
+	CHECK(a->next()->next()->get() == "I");
+	CHECK(c->prev()->get() == "I");
+	CHECK(list.front()->next()->next()->get() == "I");
+	CHECK(list.back()->prev()->get() == "I");
+}
+
+TEST_CASE("[List] Insert before null") {
+	List<String> list;
+	List<String>::Element *a = list.push_back("A");
+	List<String>::Element *b = list.push_back("B");
+	List<String>::Element *c = list.push_back("C");
+
+	list.insert_before(nullptr, "I");
+
+	CHECK(a->next()->get() == "B");
+	CHECK(b->get() == "B");
+	CHECK(c->prev()->prev()->get() == "A");
+	CHECK(list.front()->next()->get() == "B");
+	CHECK(list.back()->prev()->prev()->get() == "B");
+	CHECK(list.back()->get() == "I");
+}
+
+TEST_CASE("[List] Insert after null") {
+	List<String> list;
+	List<String>::Element *a = list.push_back("A");
+	List<String>::Element *b = list.push_back("B");
+	List<String>::Element *c = list.push_back("C");
+
+	list.insert_after(nullptr, "I");
+
+	CHECK(a->next()->get() == "B");
+	CHECK(b->get() == "B");
+	CHECK(c->prev()->prev()->get() == "A");
+	CHECK(list.front()->next()->get() == "B");
+	CHECK(list.back()->prev()->prev()->get() == "B");
+	CHECK(list.back()->get() == "I");
+}
+
+TEST_CASE("[List] Find") {
+	List<int> list;
+	List<int>::Element *n[10];
+	// Indices match values.
+	populate_integers(list, n, 10);
+
+	for (int i = 0; i < 10; ++i) {
+		CHECK(n[i]->get() == list.find(i)->get());
+	}
+}
+
+TEST_CASE("[List] Erase (by value)") {
+	List<int> list;
+	List<int>::Element *n[10];
+	// Indices match values.
+	populate_integers(list, n, 10);
+
+	bool erased = list.erase(2);
+	CHECK(erased);
+	CHECK(list.size() == 9);
+	CHECK_MESSAGE(n[2] != nullptr,
+			"The element should still be accessible before the list is cleared.");
```

Comment:
Actually I'm not sure if that's a feature, because the element should be deleted by now, but I think that the pointer is not reset to `nullptr`, makes sense?

Reply:
Wait, isn't List contiguous? If you delete the 3rd element (value 2), `n[2]` should still be valid and point to 3 no?

Reply:
I guess the issue is that you're still using the old iterator. Maybe it should be invalidated somehow if you erase an element?

If you look at
```
        T &operator[](int p_index) {
                CRASH_BAD_INDEX(p_index, size());
...
```
I'd assume that if you try to read `n[9]` after erasing the value 2, you'll get a crash, while `n[8]->next()->get()` might actually be 10.

Reply:
Note that `n` is just an array of elements which is populated for testing purposes, not a list.

So, I think the pointer should still be valid and the `CHECK_MESSAGE` should be removed, but when checking for the actual value:
```
CHECK(n[2]->get() == 2);
```
it would still hold the erased value.

I've done some changes to `List::_Data::erase()` locally by resetting the memory to zero after delete:
```
			memdelete_allocator<Element, A>(const_cast<Element *>(p_I));
			zeromem(p_I, sizeof(Element));
```
And the above assertion fails now.

Doing so would impact performance I guess, and that's relying on undefined behavior anyway.

So, I've added a comment for that in the test case.